src/runtime/profiler.cpp

```diff
@@ -45,7 +52,16 @@ WEAK halide_profiler_pipeline_stats *find_or_create_pipeline(const char *pipelin
     p->num_allocs = 0;
     p->active_threads_numerator = 0;
     p->active_threads_denominator = 0;
-    p->funcs = (halide_profiler_func_stats *)malloc(num_funcs * sizeof(halide_profiler_func_stats));
+    uint64_t bytes_per_func = sizeof(halide_profiler_func_stats);
+    #ifdef HARDWARE_COUNTERS
+    // Also allocate space to track the hardware counters
+    bytes_per_func += sizeof(halide_profiler_hardware_counter) * halide_profiler_hardware_counter_end;
+    #endif
+    p->funcs = (halide_profiler_func_stats *)malloc(num_funcs * bytes_per_func);
+    #ifdef HARDWARE_COUNTERS
+    halide_profiler_hardware_counter *counters = (halide_profiler_hardware_counter *)(p->funcs + num_funcs);
+    memset(counters, 0, num_funcs * sizeof(halide_profiler_hardware_counter) * halide_profiler_hardware_counter_end);
```

Comment:
I *think* this is correct, but it's pretty hard to read (this line and the few leading to it). I actually thought there was a bug and had a comment ready to post, but eventually realized my reading was wrong.